connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ml/MLHandler.scala

```diff
@@ -0,0 +1,231 @@
+        val stageProto = saveStageProto.getStage
+
+        stageProto.getType match {
+          case proto.MlStage.StageType.ESTIMATOR =>
+            val name = stageProto.getName
+            val algo = AlgorithmRegistry.get(name)
+            val estimator = algo.initiateEstimator(stageProto.getUid)
+            MLUtils.setInstanceParams(estimator, stageProto.getParams)
+            algo.saveEstimator(
+              estimator,
+              saveStageProto.getPath,
+              saveStageProto.getOverwrite,
+              saveStageProto.getOptionsMap.asScala.toMap)
+            proto.MlCommandResponse
+              .newBuilder()
+              .setLiteral(LiteralValueProtoConverter.toLiteralProto(null))
+              .build()
+
+          case _ =>
+            throw new UnsupportedOperationException()
+        }
+
+      case proto.MlCommand.MlCommandTypeCase.COPY_MODEL =>
+        val copyModelProto = mlCommand.getCopyModel
+        val modelEntry = sessionHolder.mlCache.modelCache.get(copyModelProto.getModelRefId)
+        val model = modelEntry._1
+        val algo = modelEntry._2
+        val copiedModel = model.copy(ParamMap.empty).asInstanceOf[Model[_]]
+        val refId = sessionHolder.mlCache.modelCache.register(copiedModel, algo)
+        proto.MlCommandResponse
+          .newBuilder()
+          .setLiteral(proto.Expression.Literal.newBuilder().setLong(refId))
+          .build()
+
+      case proto.MlCommand.MlCommandTypeCase.DELETE_MODEL =>
+        val modelRefId = mlCommand.getDeleteModel.getModelRefId
+        sessionHolder.mlCache.modelCache.remove(modelRefId)
+        proto.MlCommandResponse
+          .newBuilder()
+          .setLiteral(LiteralValueProtoConverter.toLiteralProto(null))
+          .build()
+
+      case _ =>
+        throw new IllegalArgumentException()
+    }
+  }
+
+  def transformMLRelation(
+      mlRelationProto: proto.MlRelation,
+      sessionHolder: SessionHolder): DataFrame = {
+    mlRelationProto.getMlRelationTypeCase match {
+      case proto.MlRelation.MlRelationTypeCase.MODEL_TRANSFORM =>
+        val modelTransformRelationProto = mlRelationProto.getModelTransform
+        val (model, _) =
+          sessionHolder.mlCache.modelCache.get(modelTransformRelationProto.getModelRefId)
+        // Create a copied model to avoid concurrently modify model params.
+        val copiedModel = model.copy(ParamMap.empty).asInstanceOf[Model[_]]
+        MLUtils.setInstanceParams(copiedModel, modelTransformRelationProto.getParams)
+        val inputDF =
+          MLUtils.parseRelationProto(modelTransformRelationProto.getInput, sessionHolder)
+        copiedModel.transform(inputDF)
+
+      case proto.MlRelation.MlRelationTypeCase.MODEL_ATTR =>
+        val modelAttrProto = mlRelationProto.getModelAttr
+        val modelEntry = sessionHolder.mlCache.modelCache.get(modelAttrProto.getModelRefId)
+        val model = modelEntry._1
+        val algo = modelEntry._2
```

Comment:
```suggestion
        val (model, algo) = sessionHolder.mlCache.modelCache.get(modelAttrProto.getModelRefId)
```

Can we use unpacking?

Reply:
I tried unpacking but for this case compiling failed. Error is like:
```
inferred existential type (org.apache.spark.ml.Model[_$4], org.apache.spark.sql.connect.ml.Algorithm)( forSome { type _$4 }), which cannot be expressed by wildcards,  should be enabled
by making the implicit value scala.language.existentials visible.
```

But after adding `import scala.language.existentials` it works.